### PR TITLE
[Federation] Make federation etcd PVC size configurable

### PR DESCRIPTION
**Pull request by irfanurrehman**
This one implements one of the many TODO items pending in the previous set of kubefed PRs.
The design doc PR is at https://github.com/kubernetes/kubernetes/pull/34484

cc  @kubernetes/sig-cluster-federation  @madhusudancs

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36310)

**Release note**:

``` 
[Federation] kubefed init now has a new flag, --etcd-pv-capacity, which can be used to configure the persistent volume capacity for etcd.  
```

### CODE REVIEW

**Comment by irfanurrehman** [comment]
I signed it! However I did sign the CNCF CLA quite some time back, thinking that CNCF CLA is what is only required now.

File: federation/pkg/kubefed/init/init_test.go [hunk 1]
```diff
@@ -84,6 +85,7 @@ func TestInitFederation(t *testing.T) {
 			dnsZoneName:        "example.test.",
 			lbIP:               "10.20.30.40",
 			image:              "example.test/foo:bar",
+			etcdPVCapacity:     "10Gi",
```
Comment (nikhiljindal) [inline federation/pkg/kubefed/init/init_test.go]: Will be good to have 2 tests: one where you do not specify any value and verify that it picks up 10Gi and one where you specify something other than the default value.

**Review by madhusudancs — CHANGES_REQUESTED** [review]
@irfanurrehman thank you very much for taking this up. Code looks good to me. I agree with @nikhiljindal's comment on tests. Please make that change and we will get this in.

**Comment by madhusudancs** [comment]
@k8s-bot ok to test

**Comment by irfanurrehman** [comment]
@madhusudancs @nikhiljindal please take a look at updated tests.

**Comment by irfanurrehman** [comment]
rebased the commits of this PR on top of https://github.com/kubernetes/kubernetes/pull/36492.

File: federation/pkg/kubefed/init/init_test.go [hunk 2]
```diff
@@ -392,12 +398,20 @@ func TestCertsHTTPS(t *testing.T) {
...
-func fakeInitHostFactory(federationName, namespaceName, ip, dnsZoneName, image, dnsProvider string) (cmdutil.Factory, error) {
+func fakeInitHostFactory(federationName, namespaceName, ip, dnsZoneName, image, dnsProvider, etcdPVCapacity string) (cmdutil.Factory, error) {
 	svcName := federationName + "-apiserver"
 	svcUrlPrefix := "/api/v1/namespaces/federation-system/services"
 	credSecretName := svcName + "-credentials"
 	cmKubeconfigSecretName := federationName + "-controller-manager-kubeconfig"
-	capacity, err := resource.ParseQuantity("10Gi")
+	pvCap := ""
+
+	if "" != etcdPVCapacity {
```
Comment (madhusudancs) [inline federation/pkg/kubefed/init/init_test.go]: This is a strange way to write this. Please make this the more idiomatic `etcdPVCapacity != ""` version.

File: federation/pkg/kubefed/init/init_test.go [hunk 3]
```diff
@@ -392,12 +398,20 @@ func TestCertsHTTPS(t *testing.T) {
...
+		pvCap = etcdPVCapacity
+	} else {
```
Comment (madhusudancs) [inline federation/pkg/kubefed/init/init_test.go]: Drop the else and just hoist this outside the condition. This entire thing then looks like:

```go
pvCap := "10Gi"
if etcdPVCapacity != "" {
    pvCap = etcdPVCapacity
}
```

**Review by madhusudancs — CHANGES_REQUESTED** [review]

**Comment by madhusudancs** [comment]
@irfanurrehman Can you please rebase this on master now?

**Comment by madhusudancs** [comment]
@irfanurrehman Nevermind, the PR on which this depends (#36492) was already merged, so git/Github automatically does that in the display.

**Comment by irfanurrehman** [comment]
@madhusudancs thanks for the comments; fixed them, PTAL!

File: federation/pkg/kubefed/init/init.go [hunk 4]
```diff
@@ -353,8 +355,9 @@ func createControllerManagerKubeconfigSecret(clientset *client.Clientset, namesp
 	return util.CreateKubeconfigSecret(clientset, config, namespace, kubeconfigName, false)
 }
 
-func createPVC(clientset *client.Clientset, namespace, svcName string) (*api.PersistentVolumeClaim, error) {
-	capacity, err := resource.ParseQuantity("10Gi")
+func createPVC(clientset *client.Clientset, namespace, svcName, etcdPVCapacity string) (*api.PersistentVolumeClaim, error) {
+
```
Comment (madhusudancs) [inline federation/pkg/kubefed/init/init.go]: Remove the unnecessary blank line.

**Review by madhusudancs — APPROVED** [review]

**Comment by irfanurrehman** [comment]
@madhusudancs updated; also if you still find some nits in this PR (I skimmed through the code again), then please take a look at the fix commit (last commit) in the next PR also -> https://github.com/kubernetes/kubernetes/pull/36447

additionally do i need to manually add the new flag names to hack/verify-flags/known-flags.txt ( I see one jenkins build failure repeating for the same), or is there an automated way!
Thanks for all the help!

**Comment by madhusudancs** [comment]
@irfanurrehman there is no automated way. Run ` hack/verify-flags/known-flags.txt` locally and it tells you what to add where.

**Comment by madhusudancs** [comment]
LGTM! Thanks for your work @irfanurrehman

**Comment by madhusudancs** [comment]
@irfanurrehman could you please add a release note for this one too?

**Comment by irfanurrehman** [comment]
updated hack/verify-flags/known-flags.txt and release note added.
@madhusudancs, you may please update the labels appropriately now.
also, if you find the title more suitable for the rel notes, u can probably remove the one that i did add and put a label "release-note" to use the title.

**Comment by madhusudancs** [comment]
@irfanurrehman can you please change the tone of the release note to be more informative. Think how you would want to read this if you were a user learning about the changes in the release.

Example:

"""
[Federation] `kubefed init` learned a new flag, `--etcd-pv-capacity`, which can be used to configure the persistent volume capacity for etcd. 
"""

Or something like that.

**Comment by irfanurrehman** [comment]
@madhusudancs updated as per your suggestions.

**Comment by madhusudancs** [comment]
LGTM!

**Pull request merged**